service/orchestrator/certificates.go

```diff
@@ -0,0 +1,189 @@
+package orchestrator
+
+import (
+	"context"
+	"errors"
+
+	"clouditor.io/clouditor/api/orchestrator"
+	"clouditor.io/clouditor/internal/logging"
+	"clouditor.io/clouditor/persistence"
+	"clouditor.io/clouditor/service"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+// ErrCertificationNotFound indicates the certification was not found
+var ErrCertificationNotFound = status.Error(codes.NotFound, "certificate not found")
+
+// CreateCertificate implements method for creating a new certificate
+func (svc *Service) CreateCertificate(ctx context.Context, req *orchestrator.CreateCertificateRequest) (
+	res *orchestrator.Certificate, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		return
+	}
+
+	// Check if client is allowed to access the corresponding cloud service (targeted in the certificate)
+	if !svc.authz.CheckAccess(ctx, service.AccessCreate, req) {
+		err = service.ErrPermissionDenied
+		return
+	}
+
+	// Persist the new certificate in our database
+	err = svc.storage.Create(req.Certificate)
+	if err != nil {
+		err = status.Errorf(codes.Internal, "could not add certificate to the database: %v", err)
+		return
+	}
+
+	logging.LogRequest(log, logrus.DebugLevel, logging.Create, req)
+
+	// Return certificate
+	res = req.Certificate
+	return
+}
+
+// GetCertificate implements method for getting a certificate, e.g. to show its state in the UI
+func (svc *Service) GetCertificate(ctx context.Context, req *orchestrator.GetCertificateRequest) (
+	res *orchestrator.Certificate, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		return
+	}
+
+	res = new(orchestrator.Certificate)
+	err = svc.storage.Get(res, "Id = ?", req.CertificateId)
+	if errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, ErrCertificationNotFound
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+
+	// Check if client is allowed to access the corresponding cloud service (targeted in the certificate)
+	all, allowed := svc.authz.AllowedCloudServices(ctx)
```

Comment:
It is a little bit inconsistent that for `Get` we first retrieve the object and then to an authorisation check and for `List` and `Remove` we append the cloud service ID into the query (which I would actually prefer).

I know that in other Get-methods we also do it like this but maybe we should also change it there.

Reply:
The problem is that we then cannot return a permission denied error since we cannot differentiate between the response being empty because it was not found OR because the user actually was unauthorized